OSpackages.sh

```diff
@@ -0,0 +1,399 @@
+      rpm --qf '%{name}\n' -qf  /etc/yum.repos.d/* 2>&1 | grep -v "not owned" | sort -u ; \
+      echo yum-utils
+   )
+
+   # Ensure DNS lookups work in chroot-dev
+   if [[ ! -e ${CHROOTMNT}/etc/resolv.conf ]]
+   then
+      err_exit "Installing ${CHROOTMNT}/etc/resolv.conf..." NONE
+      install -Dm 000644 /etc/resolv.conf "${CHROOTMNT}/etc/resolv.conf"
+   fi
+
+   # Ensure etc/rc.d/init.d exists in chroot-dev
+   if [[ ! -e ${CHROOTMNT}/etc/rc.d/init.d ]]
+   then
+      install -dDm 000755 "${CHROOTMNT}/etc/rc.d/init.d"
+   fi
+
+   # Ensure etc/init.d exists in chroot-dev
+   if [[ ! -e ${CHROOTMNT}/etc/init.d ]]
+   then
+      ln -t "${CHROOTMNT}/etc" -s ./rc.d/init.d
+   fi
+
+   # Clean out stale RPMs
+   if [[ $( stat /tmp/*.rpm > /dev/null 2>&1 )$? -eq 0 ]]
+   then
+      err_exit "Cleaning out stale RPMs..." NONE
+      rm -f /tmp/*.rpm || \
+        err_exit "Failed cleaning out stale RPMs"
+   fi
+
+   # Stage our base RPMs
+   yumdownloader --destdir=/tmp "${BASEPKGS[@]}"
+   if [[ ${REPORPMS:-} != '' ]]
+   then
+      FetchCustomRepos
+   fi
+
+   # Initialize RPM db in chroot-dev
+   err_exit "Initializing RPM db..." NONE
+   rpm --root "${CHROOTMNT}" --initdb || \
+     err_exit "Failed initializing RPM db"
+
+   # Install staged RPMs
+   err_exit "Installing staged RPMs..." NONE
+   rpm --root "${CHROOTMNT}" -ivh --nodeps /tmp/*.rpm || \
+     err_exit "Failed installing staged RPMs"
+
+   # Install dependences for base RPMs
+   err_exit "Installing base RPM's dependences..." NONE
+   yum --disablerepo="*" --enablerepo="${OSREPOS}" \
+      --installroot="${CHROOTMNT}" -y reinstall "${BASEPKGS[@]}" || \
+     err_exit "Failed installing base RPM's dependences"
+
+   # Ensure yum-utils are installed in chroot-dev
+   err_exit "Ensuring yum-utils are installed..." NONE
+   yum --disablerepo="*" --enablerepo="${OSREPOS}" \
+      --installroot="${CHROOTMNT}" -y install yum-utils || \
+     err_exit "Failed installing yum-utils"
+
+}
+
+# Install selected package-set into chroot-dev
+function MainInstall {
+   local YUMCMD
+
+   YUMCMD="yum --nogpgcheck --installroot=${CHROOTMNT} "
+   YUMCMD+="--disablerepo=* --enablerepo=${OSREPOS} install -y "
+
+   # Stub...
+   echo "${RPMFILE}" > /dev/null 2>&1
+
+   # Expand the "core" RPM group and store as array
+   mapfile -t INLCLUDEPKGS < <(
+      yum groupinfo "${RPMGRP}" 2>&1 | \
+      sed -n '/Mandatory/,/Optional Packages:/p' | \
+      sed -e '/^ [A-Z]/d' -e 's/^[[:space:]]*[-=+[:space:]]//'
+   )
+
+   # Add extra packages to include-list (array)
+   INLCLUDEPKGS+=( "${INLCLUDEPKGS[@]}" "${MINXTRAPKGS[@]}" )
+
+   # Remove excluded packages from include-list
+   for EXCLUDE in ${EXCLUDEPKGS[*]}
+   do
+       INLCLUDEPKGS=( "${INLCLUDEPKGS[@]//*${EXCLUDE}*}" )
+   done
+
+   # Install packages
+   YUMCMD+="$( IFS=' ' ; echo "${MINXTRAPKGS[*]}" ) " \
+   YUMCMD+="@${RPMGRP}"
+   ${YUMCMD} -x "$( IFS=',' ; echo "${EXCLUDEPKGS[*]}" )"
```

Comment:
i'm not quite following the logic here. how is this installing INCLUDEPKGS? or is this approach just [totally different from amigen7](https://github.com/plus3it/AMIgen7/blob/master/ChrootBuild.sh#L422)?

Reply:
EL8 uses `dnf` (which is a wrapper for `yum4`). The wrapping/bridgings is a bit... Flaky: 1. "order counts"; 2. negative-specification method changes.

Reply:
Originally, was going to make the whole thing be something like:

~~~
yum install <MINXTRAPKGS> <EXCLUDEPKGS> @${RPMGRP}
~~~

To make error messages around not understanding the exclusion-groups go away. DNF appears to get _really_ confused by ordering – it would toss out the message:

~~~
No match for group package "iwl5150-firmware"
No match for group package "iwl6050-firmware"
No match for group package "iwl7260-firmware"
No match for group package "iprutils"
No match for group package "iwl105-firmware"
No match for group package "iwl6000-firmware"
No match for group package "iwl2000-firmware"
No match for group package "iwl100-firmware"
No match for group package "biosdevname"
No match for group package "iwl6000g2a-firmware"
No match for group package "iwl3945-firmware"
No match for group package "iwl1000-firmware"
No match for group package "iwl3160-firmware"
No match for group package "iwl135-firmware"
No match for group package "iwl2030-firmware"
No match for group package "iwl4965-firmware"
No match for group package "iwl5000-firmware"
Dependencies resolved.
~~~

If group-specification came before the exclusion-list. At any rate, I temporarily changed the order to make the messages go away. That said, discovered that the resultant AMIs _actually had_ the firmware excluded (guessing that even though `dnf` bitches, once it passes everything through to `yum4`, the right things get handled) regardless of whether the messages are provoked or not:

~~~
# chroot /mnt/ec2-root rpm -qa \*-firmware
linux-firmware-20190516-94.git711d3297.el8.noarch
~~~

Ultimately decided to use group-specification for the install – vice the expansion of the group – and to just use the group-expansion for the [post-install check-section](# chroot /mnt/ec2-root rpm -qa \*-firmware
linux-firmware-20190516-94.git711d3297.el8.noarch
) because it made installing "not core" groups a bit more reliable without sacrificing the reliability of the check-section.

Reply:
i think i'm tracking. this is not offering the "manifest" approach currently?

Reply:
Currently, no. When I add manifest support, the invocation will most likely be "if group, do...; if manifest do...".

The changes that I made with respect to the check method means that the group method should work more broadly (except where the repo's borked) ...but I won't be able to fully validate the manifest-method until I have access to broken repos, again.

Mostly, I wanted to get a "first pass"  out with the "when everything's like it ought to be" codeup so I can get to work on spel (etc.), then circle back for the "weird" cases.